### PR TITLE
fix(ai): neutralize header-form leaked control tokens (gpt-5.6 invalid_prompt)

### LINKED ISSUE #2267: invalid_prompt still permanently wedges gpt-5.6 sessions after #2144/#2192/#2197 (header-form control tokens bypass the sanitizer)
## Summary

After #2144, #2192, and #2197, `Request blocked (code=invalid_prompt)` **still permanently wedges** `openai-codex` (gpt-5.6-*) sessions.

All three fixes funnel through a single regex — `RESERVED_CONTROL_TOKEN_RE` in `packages/ai/src/utils.ts` — that only matches the simple `<|ident|>` form. A leaked **header-form** Harmony marker whose body contains a space / `=` / `.` — e.g. `<|assistant to=functions.bash|>` — is **not** neutralized on any path (replay-history sanitizer, request-boundary deep walk, or compaction), so it reaches gpt-5.6 verbatim and re-poisons every subsequent turn.

Reproduced on `gjc 0.10.2`, which already ships #2144 + #2192 + #2197. Model: `gpt-5.6-sol` (`openai-codex-responses`).

## Impact

- Same permanent-wedge signature #2144/#2192 describe: instant `invalid_prompt`, **0 tokens**, non-recoverable because the tainted item is replayed each turn; even a benign follow-up user message fails (the tainted history is what is rejected, not the new input).
- **Self-propagating.** A session that merely *quotes* the marker — e.g. while debugging a wedged session, or reading a poisoned transcript into context — poisons its own history and wedges too. Observed exactly this: a debugging session investigating a wedged session died the same way immediately after it echoed the marker into its own transcript.
- Per #2192, retrying against tainted history can escalate to **account-level blocking**.

## Root cause

`packages/ai/src/utils.ts`:

```ts
const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[A-Za-z0-9_]{1,32}\|>)/g;
```

The `[A-Za-z0-9_]{1,32}` body class only matches simple identifiers. A degenerate Harmony role/recipient header rendered as a literal `<|role to=functions.name|>` contains spaces, `=`, and `.`, so the lookahead fails and the marker is left intact. Because `neutralizeReservedControlTokens` is the single primitive behind `neutralizeResponsesInputControlTokens` (request boundary, #2192), the replay-history sanitizer (#2144), and the compaction sanitizers (#2197), one gap defeats all of them at once.

## Reproduction (deterministic, exact current function body)

```js
const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[A-Za-z0-9_]{1,32}\|>)/g;
const n = t => t.includes("<|") ? t.replace(RESERVED_CONTROL_TOKEN_RE, "<\u200b|") : t;

n("<|channel|>analysis")     // neutralized, no raw "<|" remains          OK
n('<|assistant to=functions.bash|>')                   // UNCHANGED, raw "<|" survives -> reaches gpt-5.6   BUG
```

(#2192 already established that a leaked `<|channel|>analysis` literal is sufficient to trigger `Request blocked` on all `gpt-5.6-*` models; the header form is the same class of marker, just outside the current regex.)

## Proposed fix

Broaden the body class to any bounded run of non-`<>|` / non-newline chars, keeping the idempotent zero-width-space insertion:

```diff
-const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[A-Za-z0-9_]{1,32}\|>)/g;
+const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[^<>|\n]{1,64}\|>)/g;
```

Verified against an isolated harness that mirrors the source function (14/14):

- header form now neutralized;
- every originally-covered simple form (`<|channel|>`, `<|message|>`, `<|call|>`, `<|constrain|>`, `<|recipient|>`, `<|content|>`, `<|end|>`, `<|im_start|>`) still neutralized — no regression;
- the real leaked dump fully neutralized;
- idempotent (second pass is a no-op);
- ordinary text (`a < b and c > d, ratio a|b`) untouched;
- a >64-char run between `<|` and `|>` is left as-is (no runaway match).

Regression test added: `packages/ai/test/control-token-header-form-repro.test.ts`.

<details><summary>Full patch (<code>packages/ai/src/utils.ts</code> + regression test)</summary>

```diff
diff --git a/packages/ai/src/utils.ts b/packages/ai/src/utils.ts
index d17bf44..8bbdb0c 100644
--- a/packages/ai/src/utils.ts
+++ b/packages/ai/src/utils.ts
@@ -119,7 +119,7 @@ export function sanitizeOpenAIResponsesHistoryItemsForReplay(items: Array<Record
 		return sanitized ? [sanitized] : [];
 	});
 }
-const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[A-Za-z0-9_]{1,32}\|>)/g;
+const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[^<>|\n]{1,64}\|>)/g;
 /**
  * Neutralize leaked OpenAI Harmony / control tokens (`<|channel|>`, `<|message|>`,
  * `<|call|>`, `<|constrain|>`, `<|recipient|>`, `<|content|>`, ...) in replayed
@@ -130,6 +130,12 @@ const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[A-Za-z0-9_]{1,32}\|>)/g;
  * the offending item is re-sent on each turn. Insert a zero-width space after `<`
  * so the delimiter can no longer be tokenized as a reserved control token while the
  * text stays human-readable.
+ *
+ * The match covers both the simple `<|ident|>` form and the header form whose body
+ * carries a recipient/space/dot (e.g. `<|assistant to=functions.bash|>`); the latter
+ * survived the earlier `[A-Za-z0-9_]`-only pattern and kept bricking sessions. Any
+ * `<|…|>` run of up to 64 non-`<>|`/non-newline chars is neutralized; the inserted
+ * zero-width space makes the rewrite idempotent (`<\u200b|` no longer matches `<|`).
  */
 export function neutralizeReservedControlTokens(text: string): string {
 	if (!text.includes("<|")) return text;
diff --git a/packages/ai/test/control-token-header-form-repro.test.ts b/packages/ai/test/control-token-header-form-repro.test.ts
new file mode 100644
index 0000000..24a3f57
--- /dev/null
+++ b/packages/ai/test/control-token-header-form-repro.test.ts
@@ -0,0 +1,51 @@
+import { describe, expect, it } from "bun:test";
+import { neutralizeResponsesInputControlTokens, neutralizeReservedControlTokens } from "../src/utils";
+
+// Follow-up to #2144/#2192/#2197. The reserved-control-token sanitizer only
+// neutralized simple `<|ident|>` markers, so a leaked *header-form* Harmony
+// marker whose body contains a space / `=` / `.` (e.g. `<|assistant to=functions.bash|>`)
+// survived replay and kept the gpt-5.6 / Codex endpoint rejecting every turn with
+// `Request blocked (code=invalid_prompt)`, permanently wedging the session.
+const HEADER_FORM = "<|assistant to=functions.bash|>";
+
+function hasRawControlToken(text: string): boolean {
+	return text.includes("<|");
+}
+
+describe("neutralizeReservedControlTokens header-form markers", () => {
+	it("neutralizes a leaked role-header marker with spaces/=/. in the body", () => {
+		const out = neutralizeReservedControlTokens(`before ${HEADER_FORM} after`);
+		expect(hasRawControlToken(out)).toBe(false);
+		expect(out).toContain("<\u200b|assistant to=functions.bash|>");
+	});
+
+	it("still neutralizes simple markers and mixed dumps", () => {
+		const dump = "Not blocked.<|assistant to=functions.bash|>run<|channel|>analysis<|end|>";
+		const out = neutralizeReservedControlTokens(dump);
+		expect(hasRawControlToken(out)).toBe(false);
+	});
+
+	it("is idempotent: already-neutralized text is left unchanged", () => {
+		const once = neutralizeReservedControlTokens(`x ${HEADER_FORM} y`);
+		const twice = neutralizeReservedControlTokens(once);
+		expect(twice).toBe(once);
+	});
+
+	it("neutralizes header-form markers nested anywhere in the responses input", () => {
+		const input = [
+			{
+				type: "message",
+				role: "assistant",
+				content: [{ type: "output_text", text: `plan ${HEADER_FORM}` }],
+			},
+			{ type: "function_call_output", output: `tool said ${HEADER_FORM}` },
+		];
+		const out = neutralizeResponsesInputControlTokens(input);
+		expect(hasRawControlToken(JSON.stringify(out))).toBe(false);
+	});
+
+	it("does not touch ordinary text without control-token markers", () => {
+		const plain = "a < b and c > d, ratio a|b";
+		expect(neutralizeReservedControlTokens(plain)).toBe(plain);
+	});
+});

```

</details>

## Still open: the deferred follow-up from #2144

#2144 explicitly deferred (and it was never filed) a **session-level circuit-breaker** so a hard `invalid_prompt`/blocked turn is not re-attempted against identical tainted history, plus an **operator command to repair/trim poisoned history**. This incident is the exact motivation: widening the regex stops *new* poisoning, but it does not repair sessions already wedged, nor does it stop the tight instant-retry loop (observed ~14+ consecutive 0-token failures). Today the only recovery is to hand-edit the session `.jsonl` and neutralize the markers.

## Environment

- `gjc` 0.10.2 (`@gajae-code/coding-agent`), `@gajae-code/ai` at the same release
- model `gpt-5.6-sol`, provider `openai-codex` (`openai-codex-responses` transport)

### PR DESCRIPTION
**Pull request by hoon-ch**
## Problem

After #2144, #2192, and #2197, `Request blocked (code=invalid_prompt)` **still permanently wedges** `openai-codex` (gpt-5.6-*) sessions. Reproduced on `gjc 0.10.2` (ships all three). Details in #2267.

All three prior fixes route through one regex — `RESERVED_CONTROL_TOKEN_RE` in `packages/ai/src/utils.ts` — which only matched the simple `<|ident|>` form:

```ts
const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[A-Za-z0-9_]{1,32}\|>)/g;
```

A leaked **header-form** Harmony marker whose body carries a recipient (space + `to=` + dotted tool name) — e.g. `<|assistant to=functions.bash|>` — fails the `[A-Za-z0-9_]{1,32}` lookahead, survives replay/request-boundary/compaction sanitizing, and reaches gpt-5.6 verbatim, re-poisoning every subsequent turn (instant `invalid_prompt`, 0 tokens, unrecoverable). It is also self-propagating: a session that merely *quotes* the marker while debugging poisons its own history.

## Fix

Anchor the body to a leading identifier char and drop the length cap:

```diff
-const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[A-Za-z0-9_]{1,32}\|>)/g;
+const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[A-Za-z0-9_]+[^<>|\n]*\|>)/g;
```

- **Strict superset** of the original identifier-only pattern — every marker the old regex caught still matches.
- Header forms (`<|role to=functions.name|>`) are now covered, with **no length cap**, so arbitrarily long MCP/custom tool recipients are caught.
- Requiring an identifier char immediately after `<|` leaves **non-control pipe syntax untouched** — e.g. F# `value <| f |> g` has a space after `<|`, so it never matches.

`neutralizeReservedControlTokens` is the single primitive behind `neutralizeResponsesInputControlTokens` (request boundary), the replay-history sanitizer, and the compaction sanitizers, so this one change closes the gap on every path.

## Tests

New `packages/ai/test/control-token-header-form-repro.test.ts`:

- header-form marker neutralized (no raw `<|` survives) and stays human-readable;
- simple markers and mixed dumps still neutralized (no regression vs #2144/#2192);
- idempotent; nested-anywhere in the responses `input` array; ordinary text untouched;
- **header recipient longer than the old 64-char cap** is neutralized (review follow-up);
- **F# pipe operators** (`value <| f |> g`, `xs |> List.map f <| seed`, `a <| b |> c`) left byte-identical (review follow-up).

Verified locally (darwin-arm64, bun 1.3.14):

- `bun test packages/ai/test/control-token-header-form-repro.test.ts` → **7 pass / 0 fail**
- `bun test packages/ai/test/openai-responses-history-payload.test.ts` (sanitizer integration) → **30 pass / 0 fail**
- `bun test packages/ai` (full package) → **1662 pass / 337 skip / 0 fail** (1999 tests, 199 files)
- `biome check` (changed files) → clean
- `bun --cwd=packages/ai run check:types` (`tsc --noEmit`) → clean

## Review

Both Codex review P2s (long-recipient cap, F# false positive) addressed in `e06e7df`; see replies on the review threads.

## Follow-up (not this PR)

The session-level circuit-breaker + operator "repair poisoned history" command that #2144 deferred are still unfiled and still needed — widening the regex stops *new* poisoning but does not repair already-wedged sessions or stop the tight instant-retry loop. Tracked in #2267.

Closes #2267

### CODE REVIEW

File: packages/ai/src/utils.ts [hunk 1]
```diff
@@ -119,7 +119,7 @@ export function sanitizeOpenAIResponsesHistoryItemsForReplay(items: Array<Record
 		return sanitized ? [sanitized] : [];
 	});
 }
-const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[A-Za-z0-9_]{1,32}\|>)/g;
+const RESERVED_CONTROL_TOKEN_RE = /<\|(?=[^<>|\n]{1,64}\|>)/g;
```
[the remark this answers is not in the document]
Comment (hoon-ch) [reply packages/ai/src/utils.ts]: Good catch — fixed in e06e7df. Dropped the length cap entirely: the pattern is now `/<\|(?=[A-Za-z0-9_]+[^<>|\n]*\|>)/g`, so the header body is `identifier` + arbitrarily-long `[^<>|\n]*` recipient. A `<|assistant to=functions.<96-char-name>|>` marker is now neutralized (new regression test asserts `marker.length > 64` and that no raw `<|` survives). Full `packages/ai`: 1662 pass / 0 fail.

[the remark this answers is not in the document]
Comment (hoon-ch) [reply packages/ai/src/utils.ts]: Agreed, real false positive — fixed in e06e7df. The new pattern requires an **identifier char immediately after `<|`** (`[A-Za-z0-9_]+…`), so F# `value <| f |> g` (space after `<|`) no longer matches and is left byte-identical. It's a strict superset of the original `[A-Za-z0-9_]{1,32}` pattern — every marker the old regex caught still matches — while header forms like `<|assistant to=functions.bash|>` are now covered. New regression test asserts `value <| f |> g`, `xs |> List.map f <| seed`, and `a <| b |> c` are unchanged.

**Comment by Yeachan-Heo** [comment]
REQUEST_CHANGES

Blocking issue: the widened lookahead accepts any HEADER_MARKER body that begins with an ASCII identifier character and then contains arbitrary non-delimiter, non-LF characters. That is broader than the documented header form and newly rewrites compact pipe-operator/code syntax whose body contains spaces or punctuation. The added false-positive tests cover only forms with whitespace immediately after the opener, so they do not exercise this newly admitted overlap.

Required fix: scope the extension to the actual header grammar (for example, the role plus recipient assignment shape) while retaining the existing simple-form handling and unbounded recipient length. Add negative tests for compact pipe-bearing code/operator forms and ordinary-language lookalikes, plus positive/negative coverage for Unicode, LF/CRLF boundaries, pipe-bearing arguments, repeated/idempotent application, nested replay fields, and request-boundary sanitization.

Evidence: SHA256SUMS verified; sanitized evidence raw opener count is zero; exact head 8829b2a163c4d3f942f87376013ec923dce6b1a4 targets dev 6579d9d83e55eeb2f1b760802db6437c18de60c0; REST merge state is CLEAN; hosted Dev CI run 29385893564 completed successfully at that head, including plan, state gates, native-build, AI check, CLI smoke, focused repro test, and aggregate.
—
*[repo owner's gaebal-gajae (clawdbot) 🦞]*
